napbot/extensions/music/song.py

```diff
@@ -162,9 +200,160 @@ def __str__(self):
+                    self.status = SongStatus.NOT_AVAILABLE
+                    return
+        except Exception as cacheError:
+            log.error(f"Error pruning cache: {cacheError}")
+            self.status = SongStatus.NOT_AVAILABLE
+            return
+
+        log.info(f"Downloading song {self}")
+
+        yt_opts = {'extract_flat': 'discard_in_playlist',
+                   'final_ext': 'mp3',
+                   'format': 'bestaudio/best',
+                   'fragment_retries': 10,
+                   'ignoreerrors': 'only_download',
+                   'noplaylist': True,
+                   'outtmpl': {'default': f'{folder}/%(id)s.%(ext)s'},
+                   'postprocessors': [{'format': 'lrc',
+                                       'key': 'FFmpegSubtitlesConvertor',
+                                       'when': 'before_dl'},
+                                      {'format': 'jpg',
+                                       'key': 'FFmpegThumbnailsConvertor',
+                                       'when': 'before_dl'},
+                                      {'key': 'FFmpegExtractAudio',
+                                       'nopostoverwrites': False,
+                                       'preferredcodec': 'mp3',
+                                       'preferredquality': '5'},
+                                      {'key': 'FFmpegConcat',
+                                       'only_multi_video': True,
+                                       'when': 'playlist'}],
+                   'writesubtitles': True,
+                   'subtitleslangs': ['.*orig', 'en'],
+                   'writethumbnail': True}
+        log.debug(f"Youtube options: {yt_opts}")
+        status_code = 1
+        try:
+            with yt_dlp.YoutubeDL(yt_opts) as ydl:
+                status_code = ydl.download(self.external_id)
+        except Exception as e:
+            log.error(f"Error downloading with exception {e}")
+            status_code = 1
+
+        if status_code > 0:
+            log.warn(f"Youtube download failed for {self}")
+        else:
+            log.info(f"Downloaded to {path} for {self}")
+
+
+        try:
+            with Image.open(thumbnail_path) as image, io.BytesIO() as output:
+                image.convert("RGB").save(output, format="JPEG")
+                self.set_art(output.getvalue())
+
+                self.get_subtitles(SpotifySong._find_subtitle_file(folder, self.external_id))
+        except Exception as error:
+            log.warn(f"Error setting thumbnail {thumbnail_path} and subtitles, with {error}")
+
+        self.status = SongStatus.AVAILABLE if status_code == 0 else SongStatus.NOT_FOUND
+        reinsert_queue()
+
+class SongQueue[T](asyncio.PriorityQueue[T]):
     _queue: deque[T]
+    _maxDownloadSize = config.config["music"].getint("MaxDownloadQueue", 5)
+    _download_executor = concurrent.futures.ThreadPoolExecutor(
```

Comment:
ummmmmmmmmmm

regrettably i am not super familiar with the event loop stuff so idk if i can actually understand this in my giant-ass file

Reply:
it just creates multiple downloading threads

.... i guess i should add some comments